### PR TITLE
Collision detection supports uniform scaling

### LINKED ISSUE #71633: [4.0beta13] Physics: Scaled Area3D fails to produce signal _on_body_entered when entered by RigidBody
### Godot version

4.0 beta 13 (official)

### System information

Windows 10, NVIDIA GeForce 1060 GTX 3GB, Vulkan

### Issue description

 Area3D fails to produce signal _on_body_entered when scaled and entered by RigidBody. This worked in previous versions.

### Steps to reproduce

Set up scene with Area3d and RigidBody3d so that the RigidBody will fall into the Area3D.
Connect a script with the _on_body_entered signal.

Compare outcome when Area3D is unscaled and scaled. The Area should be a lot bigger than the RigidBody.
I used a Sphere Collision for the RigidBody and Box for the Area.

(see minimal reproduction project)

### Minimal reproduction project

[area3d_test.zip](https://github.com/godotengine/godot/files/10448954/area3d_test.zip)

### PR DESCRIPTION
**Pull request by peastman**
This fixes errors that were introduced by #70660 and #70787.  You're supposed to be able to uniformly scale collision objects.  The new collision routines ignored the scaling.

In addition to these fixes, there are other problems that prevent `move_and_collide()` from working correctly with scaled objects.  This PR does not attempt to fix them.  It just fixes the new problems introduced in those two PRs.

Nonuniform scaling still doesn't work, and I don't think there are any plans to try to make it work.  That's a much harder problem.

*Physics squad edit:*
- Fixes https://github.com/godotengine/godot/issues/71633

### CODE REVIEW

**Comment by Malcolmnixon** [comment]
I tested with https://github.com/Malcolmnixon/PhysicsCollisionTest.git and I get the following results from testing:
 - Sphere/sphere handles scaling correctly
 - Sphere/box produces incorrect collision points
 - Sphere/capsule looks good (but sweep-collision detection fails almost all the time)
 - Sphere/cylinder produces incorrect collision points
 - Capsule/capsule looks good (but sweep-collision detection fails almost all the time)

File: servers/physics_3d/godot_collision_solver_3d_sat.cpp [hunk 1]
```diff
@@ -843,7 +843,7 @@ static void _collision_sphere_box(const GodotShape3D *p_a, const Transform3D &p_
 	// Find the point on the box nearest to the center of the sphere.
 
 	Vector3 center = p_transform_b.xform_inv(p_transform_a.origin);
-	Vector3 extents = box_B->get_half_extents();
+	Vector3 extents = box_B->get_half_extents() * p_transform_b.basis[0].length();
```
Comment (Malcolmnixon) [inline servers/physics_3d/godot_collision_solver_3d_sat.cpp:846]: It looks like the `p_transform_b.xform_inv()` works if the transform consists of only translation and rotation elements, if it includes scaling then the result is bogus.

To make it work I changed the two lines as follows:
```
	Vector3 center = p_transform_b.affine_inverse().xform(p_transform_a.origin);
	Vector3 extents = box_B->get_half_extents();
```

The affine_inverse() produces a inversion-transform which works for translation/rotation/scale. The result is that center is correctly translated to the box_B coordinate system.

File: servers/physics_3d/godot_collision_solver_3d_sat.cpp [hunk 2]
```diff
@@ -877,11 +878,13 @@ static void _collision_sphere_capsule(const GodotShape3D *p_a, const Transform3D
...
 
-	real_t capsule_B_radius = capsule_B->get_radius();
+	real_t scale_A = p_transform_a.basis[0].length();
+	real_t scale_B = p_transform_b.basis[0].length();
+	real_t capsule_B_radius = capsule_B->get_radius() * scale_B;
 
 	// Construct the capsule segment (ball-center to ball-center)
 	Vector3 capsule_segment[2];
-	Vector3 capsule_axis = p_transform_b.basis.get_column(1) * (capsule_B->get_height() * 0.5 - capsule_B_radius);
+	Vector3 capsule_axis = p_transform_b.basis.get_column(1) * (capsule_B->get_height() * scale_B * 0.5 - capsule_B_radius);
```
Comment (Malcolmnixon) [inline servers/physics_3d/godot_collision_solver_3d_sat.cpp:887]: This is double-applying the `p_transform_b` scale. Firstly the scaling is performed by the algebraic `scale_B` and modified `capsule_B_radius`, and then scaled again by multiplying by the scaled `p_transform_b` basis.

I got this to work by:
 1. Restoring the `capsule_B_radius` creation line to original (unscaled)
 2. Restoring the `capsule_axis` creation line to original (scaled once by `p_transform_b.basis` multiply)
 3. Adding the multiply of `capsule_B_radius * scale_B` in the `analytic_sphere_collision` call.

File: servers/physics_3d/godot_collision_solver_3d_sat.cpp [hunk 3]
```diff
@@ -905,22 +908,24 @@ static void analytic_sphere_cylinder_collision(real_t p_radius_a, real_t p_radiu
 
 	Vector3 center = p_transform_b.xform_inv(p_transform_a.origin);
```
Comment (Malcolmnixon) [inline servers/physics_3d/godot_collision_solver_3d_sat.cpp:909]: This one suffers from a similar problem to the sphere/box in that the `p_transform_b.xform_inv()` doesn't work correctly when the transform is scaled. I got this to pass testing by:
 1. Using `center = p_transform_b.affine_inverse().xform(p_transform_a.origin)` so center is completely in the cylinders coordinate space without need for scaling.
 2. Remove the radius and half-height scaling by `scale_B` as center is now in the cylinders frame of reference.

The rest of the code works fine - the `nearest = p_transform_b.xform(nearest)` now correctly translates from cylinder-centric coordinates back to global coordinates including re-scaling back.

File: servers/physics_3d/godot_collision_solver_3d_sat.cpp [hunk 4]
```diff
@@ -1632,14 +1637,16 @@ static void _collision_capsule_capsule(const GodotShape3D *p_a, const Transform3
...
+	real_t scale_B = p_transform_b.basis[0].length();
+	real_t capsule_A_radius = capsule_A->get_radius() * scale_A;
+	real_t capsule_B_radius = capsule_B->get_radius() * scale_B;
 
 	// Get the closest points between the capsule segments
 	Vector3 capsule_A_closest;
 	Vector3 capsule_B_closest;
-	Vector3 capsule_A_axis = p_transform_a.basis.get_column(1) * (capsule_A->get_height() * 0.5 - capsule_A_radius);
-	Vector3 capsule_B_axis = p_transform_b.basis.get_column(1) * (capsule_B->get_height() * 0.5 - capsule_B_radius);
+	Vector3 capsule_A_axis = p_transform_a.basis.get_column(1) * (capsule_A->get_height() * scale_A * 0.5 - capsule_A_radius);
```
Comment (Malcolmnixon) [inline servers/physics_3d/godot_collision_solver_3d_sat.cpp:1648]: This has the same problem as the sphere-capsule in that the transforms are being double-applied by both multiply by `scale_[A/B]`, and then multiplying by the scaled `p_transform_[a/b].basis`.

I got this to work by:
1. Removing the scaling of `capsule_[A/B]_radius` by the associated `scale_[A/B]`
2. Removing the scaling of `capsule_[A/B]->get_height()` by the associated `scale_[A/B]`
3. This causes the `capsule_[A/B]_axis` to be scaled only once so it's now in correct global coordinates
4. Adding the multiplies by `scale_[A/B]` into the `analytic_sphere_collision` call

**Comment by peastman** [comment]
Thanks for the quick analysis and debugging!  I updated it with the changes you suggested.

**Comment by Malcolmnixon** [comment]
This is looking very good. With this PR I get the following collision results (with scaling on the shapes):
|          | Sphere | Capsule | Cylinder | Box | Convex | Concave | Boundary |
|----------|--------|---------|----------|-----|--------|---------|----------|
| Sphere   | OK     | OK      | OK       | OK  | OK     | OK      | OK       |
| Capsule  | OK     | OK      | OK       | OK  | OK     | OK      | OK       |
| Cylinder | OK     | OK      | 1        | OK  | 1      | OK      | OK       |
| Box      | OK     | OK      | OK       | OK  | OK     | 2       | OK       |
| Convex   | OK     | 3       | 1        | OK  | OK     | OK      | OK       |
| Concave  | 4      | 4       | 4        | 4   | 1      | /       | /        |

1. Numerous sweep misses
2. Normal glitch where box-corner hits concave inner crease
3. Collision points show significant positional error
4. Rare sweep misses

**Comment by fabriceci** [comment]
@Malcolmnixon In a previous table you had noted that Box/Concave was working properly, I see that this is no longer the case, is this a regression?

**Comment by peastman** [comment]
We're making good progress!  It looks like most of the remaining problems involve cylinders and concave objects.  Those are the hardest ones.

> Normal glitch where box-corner hits concave inner crease

There might not be a lot we can do about that.  A box corner doesn't have a well defined normal.  Neither does a sharp edge.  The only way to define a normal from the combination of the two is by penetration direction, which can be noisy.

**Comment by Malcolmnixon** [comment]
> @Malcolmnixon In a previous table you had noted that Box/Concave was working properly, I see that this is no longer the case, is this a regression?

I don't think it's a regression - This time through I set the box tumbling and watched it for a while - I also extended the surface normal shape so it was easier to see glitches. The normal only goes wonky for less than half a second every minute, and it seems to be when the corner of the cube rolls over an edge of the concave shape. I probably just didn't notice it before.

**Comment by Malcolmnixon** [comment]
> t we can do about that. A box corner doesn't have a well defined normal. Neither does a sharp edge. The only way to define a normal from the combination of the two is by penetration direction, which can be noisy.

I agree, it's not well defined, and it would probably take significant effort to try and clamp the normal so its between the normal s of the two neighboring faces of the edge.

**Review by rburing — APPROVED** [review]
Thanks, both!

**Comment by akien-mga** [comment]
Thanks!

**Comment by peastman** [comment]
It looks like the inaccurate positions with capsule-convex are more SAT precision problems.  The separation axis has a y component that is just barely less than 1, and tiny x and y components.  When they get multiplied by 1000, you get very inaccurate results.

**Comment by peastman** [comment]
Actually, I think there's something more than that wrong with it.  I'll see if I can figure it out.

**Pull request merged**